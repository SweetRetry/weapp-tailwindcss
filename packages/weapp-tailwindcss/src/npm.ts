import fetch from 'npm-registry-fetch'

export function fetchPackage(packageName: string) {
  return fetch.json(`/${packageName}`) as Promise<{
    'dist-tags': {
      latest: string
    }
    'versions': Record<string, unknown>
  }>
}

export async function getLatestVersion(packageName: string) {
  const response = await fetchPackage(packageName)
  return response['dist-tags'].latest
}

export async function getLatestVersionInRange(packageName: string, versionRange: string) {
  const response = await fetchPackage(packageName)
  const versions = Object.keys(response.versions)
  // 过滤出符合指定版本范围的版本
  const filteredVersions = versions.filter(version => version.startsWith(versionRange))
  // 找到符合条件的最新版本
  return filteredVersions[filteredVersions.length - 1]
}

// tailwindcss postcss autoprefixer weapp-tailwindcss
export const defaultDevDeps = {
  'tailwindcss': '3',
  'postcss': '8',
  'autoprefixer': '10',
  'weapp-tailwindcss': '3',
}

export async function getDevDepsVersions() {
  return Object.fromEntries(await Promise.all(
    Object.entries(defaultDevDeps).map(
      async (x) => {
        return [x[0], `^${await getLatestVersionInRange(...x)}`]
      },
    ),
  )) as typeof defaultDevDeps
}
