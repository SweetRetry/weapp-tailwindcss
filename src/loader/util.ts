import fs from 'fs'
import path from 'path'

export function mkfileSync (filename: string, content: string) {
  try {
    const dirname = path.dirname(filename)
    if (!fs.existsSync(dirname)) {
      fs.mkdirSync(dirname, {
        recursive: true
      })
    }
    fs.writeFileSync(filename, content, 'utf-8')
  } catch (error) {
    console.error(error)
  }
}
