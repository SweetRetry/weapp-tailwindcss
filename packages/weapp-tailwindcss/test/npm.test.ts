import { getDevDepsVersions, getLatestVersion, getLatestVersionInRange } from '@/npm'
import CI from 'ci-info'
import { fetchOptions } from './util'

describe.skipIf(CI.isCI)('getLatestVersion', () => {
  it.skip('tailwindcss', async () => {
    let version = await getLatestVersion('tailwindcss', fetchOptions)
    expect(version).toBe('3.4.13')
    version = await getLatestVersionInRange('tailwindcss', '3', fetchOptions)
    expect(version).toBe('3.4.13')
    version = await getLatestVersionInRange('tailwindcss', '2', fetchOptions)
    expect(version).toBe('2.2.19')
    // const version2 = await getLatestVersion2('tailwindcss')
    // expect(version2).toBe('3.4.12')
  })

  it('getDevDepsVersions', async () => {
    const res = await getDevDepsVersions(fetchOptions)
    console.log(res)
  })

  it('tailwindcss@2', async () => {
    const version = await getLatestVersionInRange('tailwindcss', '2', fetchOptions)
    expect(version.startsWith('2.')).toBe(true)// .toBe('2.2.19')
  })

  it('tailwindcss@3', async () => {
    const version = await getLatestVersionInRange('tailwindcss', '3', fetchOptions)
    expect(version).toBe('3.4.13')
  })

  it('tailwindcss@latest', async () => {
    const version = await getLatestVersion('tailwindcss', fetchOptions)
    expect(version).toBe('3.4.13')
  })
})
