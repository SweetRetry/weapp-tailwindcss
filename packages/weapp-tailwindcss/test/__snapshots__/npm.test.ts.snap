// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getLatestVersion > tailwindcss@3 1`] = `"3.4.17"`;

exports[`getLatestVersion > tailwindcss@latest 1`] = `"3.4.17"`;
